zkevm-circuits/src/evm_circuit/execution/calldataload.rs

```diff
@@ -0,0 +1,470 @@
+            offset.expr(),
+            offset.expr() + calldata_offset.expr(),
+        );
+        let src_addr_end = select::expr(
+            cb.curr.state.is_root.expr(),
+            calldata_length.expr(),
+            calldata_length.expr() + calldata_offset.expr(),
+        );
+
+        cb.condition(cb.curr.state.is_root.expr(), |cb| {
+            cb.tx_context_lookup(
+                tx_id.expr(),
+                TxContextFieldTag::CallDataLength,
+                None,
+                calldata_length.expr(),
+            );
+            cb.require_equal(
+                "if is_root then calldata_offset == 0",
+                calldata_offset.expr(),
+                0.expr(),
+            );
+        });
+        cb.condition(1.expr() - cb.curr.state.is_root.expr(), |cb| {
+            cb.call_context_lookup(
+                false.expr(),
+                None,
+                CallContextFieldTag::CallDataLength,
+                calldata_length.expr(),
+            );
+            cb.call_context_lookup(
+                false.expr(),
+                None,
+                CallContextFieldTag::CallDataOffset,
+                calldata_offset.expr(),
+            );
+        });
+
+        let buffer_reader = BufferReaderGadget::construct(cb, src_addr.clone(), src_addr_end);
+
+        let mut calldata_word = (0..N_BYTES_WORD)
+            .map(|idx| {
+                // for a root call, the call data comes from tx's data field.
+                cb.condition(
+                    cb.curr.state.is_root.expr() * buffer_reader.read_flag(idx),
+                    |cb| {
+                        cb.tx_context_lookup(
+                            tx_id.expr(),
+                            TxContextFieldTag::CallData,
+                            Some(src_addr.expr() + idx.expr()),
+                            buffer_reader.byte(idx),
+                        );
+                    },
+                );
+                // for an internal call, the call data comes from memory.
+                cb.condition(
+                    (1.expr() - cb.curr.state.is_root.expr()) * buffer_reader.read_flag(idx),
+                    |cb| {
+                        cb.memory_lookup(
+                            0.expr(),
+                            src_addr.expr() + idx.expr(),
+                            buffer_reader.byte(idx),
+                        );
+                    },
+                );
+                buffer_reader.byte(idx)
+            })
+            .collect::<Vec<Expression<F>>>();
+
+        // Since the stack items are in little endian form, we reverse the bytes
+        // here.
+        calldata_word.reverse();
+
+        // Add a lookup constraint for the 32-bytes that should have been pushed
+        // to the stack.
+        let calldata_word: [Expression<F>; N_BYTES_WORD] = calldata_word.try_into().unwrap();
+        cb.stack_push(RandomLinearCombination::random_linear_combine_expr(
+            calldata_word,
```

Comment:
I wonder if this works?

```suggestion
            calldata_word.try_into().unwrap(),
```

Reply:
No this does not work. Since the compiler cannot understand the size `N` of the array.

Reply:
In theory, as per: https://doc.rust-lang.org/std/primitive.array.html#impl-TryFrom%3C%26%27_%20%5BT%5D%3E

```rust
[Expression<F>; N_BYTES_WORD]::try_from(calldata_word).unwrap()
```
should work if Expression<F> is `Copy`.

But not sure it really improves readability.